megamek/src/megamek/ai/dataset/UnitAction.java

```diff
@@ -27,80 +28,89 @@
  */
 package megamek.ai.dataset;
 
+import java.util.List;
+import java.util.stream.Stream;
+
+import megamek.ai.utility.EntityFeatureUtils;
 import megamek.client.ui.SharedUtility;
-import megamek.common.Coords;
+import megamek.common.Compute;
 import megamek.common.Entity;
+import megamek.common.IAero;
 import megamek.common.MovePath;
 import megamek.common.MoveStep;
-
-import java.util.List;
+import megamek.common.UnitRole;
 
 /**
- * Represents a unit action.
- * @param id unit id
- * @param facing facing direction
- * @param fromX x coordinate of the starting position
- * @param fromY y coordinate of the starting position
- * @param toX x coordinate of the final position
- * @param toY y coordinate of the final position
- * @param hexesMoved number of hexes moved
- * @param distance distance moved
- * @param mpUsed movement points used
- * @param maxMp maximum movement points
- * @param mpP movement points percentage
- * @param heatP heat percentage
- * @param armorP armor percentage
- * @param internalP internal percentage
- * @param jumping jumping
- * @param prone prone
- * @param legal is move legal
+ * UnitAction is a record that represents the action of a unit in the game.
  * @author Luana Coppio
  */
-public record UnitAction(int id, int teamId, int playerId, String chassis, String model, int facing, int fromX, int fromY, int toX, int toY, int hexesMoved, int distance, int mpUsed,
-                         int maxMp, double mpP, double heatP, double armorP, double internalP, boolean jumping, boolean prone,
-                         boolean legal, double chanceOfFailure, List<MovePath.MoveStepType> steps, boolean bot) {
+public record UnitAction(int id, int teamId, int playerId, String chassis, String model, int facing,
+      int fromX, int fromY, int toX, int toY, int hexesMoved, int distance,
+      int mpUsed, int maxMp, double mpP, double heatP, double armorP, double internalP,
+      boolean jumping, boolean prone, boolean legal, double chanceOfFailure,
+      List<MovePath.MoveStepType> steps, boolean bot,
+      boolean hasEcm, int armor, int internal, int bv, int maxRange, int totalDamage,
+      float armorFrontP, float armorLeftP, float armorRightP, float armorBackP, UnitRole role,
+      List<Integer> weaponDamageFacingShortMediumLongRange) {
 
-    public static  UnitAction fromMovePath(MovePath movePath) {
+    /**
+     * Creates a UnitAction from a MovePath.
+     * @param movePath The MovePath to convert
+     * @return The UnitAction
+     */
+    public static UnitAction fromMovePath(MovePath movePath) {
         Entity entity = movePath.getEntity();
-        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream().map(psr -> psr.getValue() / 36d).reduce(0.0, (a, b) -> a * b);
+        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream()
```

Comment:
Why is this being pulled out here?  It should live in UnitAction if UnitAction cares about it.

Reply:
Same for `steps`, actually.

Reply:
See above; I think this can be made less boiler-platey (or that the boilerplate can be pushed down a layer).

Reply:
UnitAction does not keep the movepath, only primitive values that represents it for logging.

Those two variables vould be down there, but are up here, no big reason why.